File: solutions/clojure/knapsack/1/src/knapsack.clj
(ns knapsack)

(defn maximum-value
  "Calculates the maximum value that can be packed."
  [maximum-weight items]
  (if (empty? items) 0
    (let [num-items (count items)
          empty-table (vec (repeat (inc num-items) (vec (repeat (inc  maximum-weight) 0))))]      
      (loop [idx 0 
       wgt 1
       table empty-table]
        (if (>= idx num-items) 
          (get-in table [num-items maximum-weight])
         (if (> wgt maximum-weight)
          (recur (inc idx) 1 table)
          (let [cur (nth items idx)
                prev-val (get-in table [idx wgt])]
              (if (>= wgt (:weight cur))
                (recur idx (inc wgt) 
                       (update-in table [(inc idx) wgt]
                          (constantly
                            (max prev-val 
                               (+ (:value cur)
                                  (get-in table 
                                    [idx (- wgt (:weight cur)) ] )))))
                                   )
                (recur idx (inc wgt) 
                       (update-in table [(inc idx) wgt] 
                                  (constantly prev-val)))
              
                  ))))))))

File: solutions/julia/knapsack/1/knapsack.jl
function maximum_value(maximum_weight, items)
    0 == length(items)  && return 0
    table = zeros(Int64, length(items)+1, maximum_weight + 1)
    for (idx,itm) in enumerate(items)
        for wgt in 1:maximum_weight
            prevItemVal= table[idx,wgt+1]
            if wgt>= itm.weight
                table[idx+1,wgt+1] = max(table[idx,wgt - itm.weight+1] + itm.value, prevItemVal)
            else
                table[idx+1,wgt+1] = prevItemVal
            end
        end
    end
    table[length(items)+1, maximum_weight+1]
end